### PR TITLE
Support build constraints

### LINKED ISSUE #5561: Add constraints (and overrides?) to build environments
We should add a `--build-constraints` that apply to the build environments.

See: https://github.com/astral-sh/uv/issues/5551#issuecomment-2256055975

### PR DESCRIPTION
**Pull request by blueraft**
## Summary

Partially resolves #5561. Haven't added overrides support yet but I can add it tomorrow if the current approach for constraints is ok.

## Test Plan

`cargo test`

Manually checked trace logs after changing the constraints.

### CODE REVIEW

File: crates/uv/tests/pip_install.rs [hunk 1]
```diff
@@ -6244,3 +6244,58 @@ fn install_relocatable() -> Result<()> {
...
+        .arg("--build-constraint")
+        .arg("build_constraints.txt"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+
+    ----- stderr -----
+    Resolved 1 package in [TIME]
+    Prepared 1 package in [TIME]
+    Installed 1 package in [TIME]
```
Comment (blueraft) [inline crates/uv/tests/pip_install.rs:6295]: I wasn't sure how to check if the built constraints were being respected for the compatible versions - this just check if the package gets resolved correctly.

Comment (T-256) [reply crates/uv/tests/pip_install.rs:6295]: isn't there anything via verbose flag?

Comment (blueraft) [reply crates/uv/tests/pip_install.rs:6295]: It's there in the verbose logs, but I don't know if we should use the verbose logs for snapshot tests.

**Comment by charliermarsh** [comment]
Thanks! Will review.

File: crates/uv/src/commands/pip/compile.rs
Comment (paveldikov) [inline crates/uv/src/commands/pip/compile.rs:1]: Does `uv pip compile` now include build dependencies in its output? Or do `--build-constraints` files have to be generated separately (how?)

Comment (charliermarsh) [reply crates/uv/src/commands/pip/compile.rs:1]: I don't believe this PR adds any way to lock the build resolution, only to constrain it, but we want to add that to `uv.lock` eventually.

**Comment by hauntsaninja** [comment]
Not using `UV_CONSTRAINT` for build constraints should probably listed in the pip compatibility docs. Something like:

For constraining build time dependencies:
- pip lets you use `PIP_CONSTRAINT` environment variable to point to a constraints file. There is no CLI argument
- uv lets you use `UV_CONSTRAINT` environment variable or a `--build-constraint` CLI argument

**Comment by charliermarsh** [comment]
Makes sense, good call @hauntsaninja.

File: crates/uv/src/commands/pip/compile.rs
Comment (blueraft) [reply crates/uv/src/commands/pip/compile.rs:1]: Related issue #5190

**Review by charliermarsh — APPROVED** [review]

**Pull request merged**